source/slang/options.cpp

```diff
@@ -859,14 +859,26 @@ struct OptionsParser
 
 
 SlangResult parseOptions(
-    SlangCompileRequest*    compileRequest,
+    SlangCompileRequest*    compileRequestIn,
     int                     argc,
     char const* const*      argv)
 {
+    Slang::CompileRequest* compileRequest = (Slang::CompileRequest*) compileRequestIn;
+
     OptionsParser parser;
-    parser.compileRequest = compileRequest;
-    parser.requestImpl = (Slang::CompileRequest*) compileRequest;
-    return parser.parse(argc, argv);
+    parser.compileRequest = compileRequestIn;
+    parser.requestImpl = compileRequest;
+
+    Result res = parser.parse(argc, argv);
+
+    DiagnosticSink* sink = &compileRequest->mSink;
+    if (sink->GetErrorCount() > 0)
+    {
+        // Put the errors in the diagnostic 
+        compileRequest->mDiagnosticOutput = sink->outputBuffer.ProduceString();
```

Comment:
The only note I'd make here is that now this logic appears in more than one place, and the central reason for this PR existing is that we forgot to add the logic to a new place that needs it. This makes me wonder if there isn't a deeper problem/solution we could employ.

Right now we have this distinction between `CompileRequest::mSink` which is used for outputting diagnostic messages, and `CompileRequest::mDiagnosticOutput` which is a "dead" copy of that output generated at specific points. The latter member variable really only exists to keep the heap memory for the diagnostic output live so that the user can use the pointer they get from `spGetDiagnosticOutput`.

(One aside is that in a change many months ago I made the `StringBuilder` type just a convenience subtype of `String`, so that `sink->outputBuffer` could in principle be referenced directly instead of making a copy. The one problem there would be if the user made a call that affected the diagnostic output while expecting the pointer they last got from `spGetDiagnostic...` to remain valid)

An alternative high-level fix here might be to see if we can copy from `sink->outputBuffer` to `mDiagnosticSink` on demand when `spGetDiagnostic...` is called, so that it will properly update for any diagnostics that were output, no matter how they were generated.

There might be some thorny issues that make that fix harder than I'm imagining, but at least it would be something where the work becomes centralized so that all API operations that might produce diagnostics could share it...